lms/djangoapps/certificates/generation.py

```diff
@@ -0,0 +1,86 @@
+"""
+Course certificate generation
+
+These methods generate course certificates (they create a new course certificate if it does not yet exist, or update the
+existing cert if it does already exist).
+
+For now, these methods deal primarily with allowlist certificates, and are part of the V2 certificates revamp.
+
+These method should be called from tasks.
+"""
+
+import logging
+import random
+from uuid import uuid4
+
+import six
+from capa.xqueue_interface import make_hashkey
+from xmodule.modulestore.django import modulestore
+
+from common.djangoapps.student.models import CourseEnrollment, UserProfile
+from lms.djangoapps.certificates.models import CertificateStatuses, GeneratedCertificate
+from lms.djangoapps.certificates.utils import emit_certificate_event
+from lms.djangoapps.grades.api import CourseGradeFactory
+
+log = logging.getLogger(__name__)
+
+
+def generate_allowlist_certificate(user, course_key):
+    """
+    Generate an allowlist certificate for this user, in this course run. This method should be called from a task.
+    """
+    cert = _generate_certificate(user, course_key)
+
+    if CertificateStatuses.is_passing_status(cert.status):
+        # Emit a certificate event. Note that the two options for generation_mode are "self" (implying the user
+        # generated the cert themself) and "batch" for everything else.
+        event_data = {
+            'user_id': user.id,
+            'course_id': six.text_type(course_key),
+            'certificate_id': cert.verify_uuid,
+            'enrollment_mode': cert.mode,
+            'generation_mode': 'batch'
```

Comment:
Should the generation_mode here be hardcoded to `batch` if `self` is an option? Do all Certificates created by the Allowlist be considered the `batch` generation mode?

Reply:
For now (and in the near future), these certs won't be hooked up to generation by an end user.